Uniquify requisite_in'd states checks.

gpernot-celeste opened the pull request:
Fixes #68408

Introduce `_add_to_extend` for small factorization.

The logic here is to prevent to append dependencies that are already in  `extend[name][rstate][ind][rkey]`  :
```
for key in extend[name][rstate][ind][rkey]:
  if key.get(state) == id_:
    break
```

bdrx312 commented on salt/state.py:
```diff
@@ -1903,6 +1903,29 @@ def apply_exclude(self, high):
+        if name not in extend:
+            extend[name] = HashableOrderedDict()
+        if rstate not in extend[name]:
+            extend[name][rstate] = []
```
This should be using `state` instead of `rstate`. Naming the variables more verbosely will help some with following this.

bdrx312 reviewed (commented):
Thanks for the PR; please add tests.

bdrx312 commented:
Thanks for the PR. Nice changes.

gpernot-celeste commented:
> Thanks for the PR. Nice changes.

Thanks :)

It was buggy, but i now believe that the code is correct.

Should I rebase/squash on 3006.x ?

bdrx312 commented on salt/state.py:
```diff
@@ -1913,30 +1913,33 @@ def _add_to_extend(
 
-        if (req_states := extend[id_to_extend].get(req_state_mod_name)) is None:
-            # The requisited state is not present yet, create it and initialize it
+        if (_req_module := extend[id_to_extend].get(req_state_mod_name)) is None:
```
Oops, I overlooked that this was an old commit.

gpernot-celeste commented on salt/state.py:
```diff
@@ -1903,6 +1903,44 @@ def apply_exclude(self, high):
+            return
+
+        # Lookup req_type in _state_args
+        for _mod_req in _state_args:
```
Ok, I changed this

dwoz reviewed (changes requested):
- This needs a changelog.
- I'm super concerned that this could fix a bug and introduce others.

gpernot-celeste committed:
Uniquify requisite_in'd states checks (#68408)

Introduce ``State._add_to_extend`` and route both call sites in
``State.requisite_in`` through it. The helper dedups
``{req_state_mod_name: req_id}`` entries so a target state's requisite
list no longer accumulates duplicate references when the same
``*_in`` is declared multiple times (issue #68408).

Tests cover ``_add_to_extend`` in isolation plus both call sites of
``requisite_in`` (dict-form and list-form) for ``require_in``,
``watch_in``, ``onfail_in`` and ``onchanges_in``, including a
negative control to make sure distinct sources are NOT collapsed.

Rebased onto current 3006.x; ``HashableOrderedDict([(...)])``
positional-init usage in the test cases was switched to a small
helper because the class no longer accepts an items iterable on
3006.x.

Co-authored-by: guillaume pernot <guillaume.pernot@celeste.fr>
